### PR TITLE
	Map.of(KVKVKV)

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: javaslang/src-gen/test/java/javaslang/APITest.java [hunk 1]
```diff
@@ -729,6 +730,111 @@ public void shouldLinkedMapFromPairsReturnNotNull() {
...
     @Test
+    public void shouldCreateLinkedMapFrom1Pairs() {
+      Map<Integer, Integer> map = LinkedMap(1, 2);
+      assertThat(map.apply(1)).isEqualTo(2);
+    }
+
+    @Test
+    public void shouldCreateLinkedMapFrom2Pairs() {
+      Map<Integer, Integer> map = LinkedMap(1, 2, 2, 4);
+      assertThat(map.apply(1)).isEqualTo(2);
```
Comment (ruslansennov) [inline javaslang/src-gen/test/java/javaslang/APITest.java]: @paplorinc thank you, `apply(i)` looks better than `getOrElse(i, 0)`

File: javaslang/src/main/java/javaslang/collection/HashMap.java [hunk 2]
```diff
@@ -121,22 +121,100 @@ private HashMap(HashArrayMappedTrie<K, V> trie) {
...
+
+    /**
+     * Creates a HashMap of the given list of key-value pairs.
+     *
+     * @param <K> The key type
+     * @param <V> The value type
+     * @return A new Map containing the given entries
+     */
+    public static <K, V> HashMap<K, V> of(K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4) {
+        return new HashMap<>(HashArrayMappedTrie.<K, V> empty().put(k1, v1).put(k2, v2).put(k3, v3).put(k4, v4));
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/HashMap.java]: @ruslansennov, sorry for repeating myself, I think you could make this more compact by reusing previous factory methods, i.e.:
```java
public static <K, V> HashMap<K, V> of(K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4) {
    return of(k1, v1, k2, v2, k3, v3).put(k4, v4));
```

File: javaslang/src/main/java/javaslang/collection/LinkedHashMap.java [hunk 3]
```diff
@@ -149,6 +124,123 @@ private LinkedHashMap(Queue<Tuple2<K, V>> list, HashMap<K, V> map) {
...
+     * Creates a LinkedHashMap of the given list of key-value pairs.
+     *
+     * @param <K> The key type
+     * @param <V> The value type
+     * @return A new Map containing the given entries
+     */
+    public static <K, V> LinkedHashMap<K, V> of(K k1, V v1, K k2, V v2) {
+        final HashMap<K, V> map = HashMap.of(k1, v1, k2, v2);
+        final Queue<Tuple2<K, V>> list = Queue.of(Tuple.of(k1, v1), Tuple.of(k2, v2));
+        return new LinkedHashMap<>(list, map);
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/LinkedHashMap.java]: minor:
* consider declaring the variables in their order of usage
* it's weird to call a `Queue` `list` - though that's what the `LinkedHashMap` calls it also
* Consider using the `API` shortcuts for `HashMap.of`, `Queue.of` and `Tuple.of`, i.e. 
```java
Queue(Tuple(k1, v1), Tuple(k2, v2));
```

File: javaslang/src/main/java/javaslang/collection/TreeMap.java [hunk 4]
```diff
@@ -218,6 +290,114 @@ private TreeMap(RedBlackTree<Tuple2<K, V>> entries) {
...
+
+    /**
+     * Creates a {@code TreeMap} of the given list of key-value pairs.
+     *
+     * @param keyComparator The comparator used to sort the entries by their key.
+     * @param <K> The key type
+     * @param <V> The value type
+     * @return A new Map containing the given entries
+     */
+    public static <K extends Comparable<? super K>, V> TreeMap<K, V> of(Comparator<? super K> keyComparator, K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4, K k5, V v5, K k6, V v6) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/TreeMap.java]: in other places the params have JavaDoc also (I don't think they have value, though, but it seems inconsistent to me)

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java [hunk 5]
```diff
@@ -99,7 +99,7 @@ default int letterCount(Seq<Integer> range) {
             1_000, "thousand",
             1_000_000, "million",
             1_000_000_000, "billion"
-    );
+    ).grouped(2).toSortedMap(pair -> Tuple.of((Integer) pair.get(0), (String) pair.get(1)));
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: @ruslansennov, could you please react to my objections that you reject?
If my comments bother you and you don't find them helpful, I don't mind, but please let me know.

File: javaslang/src/main/java/javaslang/collection/HashMap.java (hunk 2 above)
Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/collection/HashMap.java]: Thank you, I will do that

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java (hunk 5 above)
Comment (ruslansennov) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: On my sight there is no difference between this solution and your proposal. My goal was to make tests passing only.

Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: Looks good to me. As Ruslan says, it works - we should spend our energy on the issues.

Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: I just wanted to draw attention to the fact that this is a real-world problem, that others will have to solve also. If the tests indicate that something is awkward, we should listen to them.

Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: Yes, but for now we do a `TreeMap.of(entries/tuples)` in a real world app for n &gt; 10.

For this test it is ok.

Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: @paplorinc please create an issue/proposal for a merge/compose API for Maps.

File: javaslang/src/main/java/javaslang/collection/HashMap.java [hunk 6]
```diff
@@ -214,7 +214,7 @@ private HashMap(HashArrayMappedTrie<K, V> trie) {
      * @return A new Map containing the given entries
      */
     public static <K, V> HashMap<K, V> of(K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4, K k5, V v5, K k6, V v6, K k7, V v7, K k8, V v8, K k9, V v9, K k10, V v10) {
-        return new HashMap<>(HashArrayMappedTrie.<K, V> empty().put(k1, v1).put(k2, v2).put(k3, v3).put(k4, v4).put(k5, v5).put(k6, v6).put(k7, v7).put(k8, v8).put(k9, v9).put(k10, v10));
+        return of(k1, v1, k2, v2, k3, v3, k4, v4, k5, v5, k6, v6, k7, v7, k8, v8, k9, v9).put(k10, v10);
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/HashMap.java]: :+1:

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java (hunk 5 above)
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: It's ok, it was just an idea, I don't mind either way :)

**Comment by danieldietrich** [comment]
@ruslansennov many thanks, this is huge!

**Pull request merged**